### PR TITLE
ARTEMIS-2641 Openwire client runs out of credits after reconnection

### PR DESCRIPTION
**Pull request by brusdev**
Clear the messagePullHandler on setting prefetchSize.

### CODE REVIEW

File: artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java [hunk 1]
```diff
@@ -380,6 +383,11 @@ public void setPrefetchSize(int prefetchSize) {
       this.prefetchSize = prefetchSize;
       this.currentWindow.set(prefetchSize);
       this.info.setPrefetchSize(prefetchSize);
+      if (this.prefetchSize == 0) {
+         messagePullHandler.compareAndSet(null, new MessagePullHandler());
```
Comment (franz1981) [inline artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java]: what happen if it won't succeed?

Comment (franz1981) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java]: @michaelandrepearce any suggestions?

Comment (michaelpearce-gain) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java]: what do you mean if it wont succeed? if its not null it means it already had a pullhandler set, no need to assign a new one.

**Comment by michaelpearce-gain** [comment]
LGTM (obviously ;) ), nice test, that was the only bit i was struggling with how to think to implement.

File: artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java (hunk 1 above)
Comment (franz1981) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java]: that can happen if: 

1. there is a racing thread doing it 
2. the value isn't null

in both cases probably logging some trace/debug is ok if we expect it to happen?

Comment (franz1981) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java]: Anyway It's fine to me; if there are no objections I'm going to merge this ;)

Comment (michaelpearce-gain) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java]: if its not null, we dont want to change it. e.g. the case there is client was already at prefetch 0 and simply sent a second update which set it to the same size, as such you want to leave the original pullhandler.

Comment (michaelpearce-gain) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java]: if you really want a log i would add it as a trace, i dont think its needed tbh i think it just clutter the code base.

Comment (brusdev) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java]: no need to set messagePullHandler value if it isn't null

Comment (franz1981) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java]: :+1:  thanks for the explanation bud :)

Comment (michaelpearce-gain) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java]: > Anyway It's fine to me; if there are no objections I'm going to merge this ;)

no objections here, please go ahead....

**Pull request closed without merging**